File: OpenBullet2.Web/Dtos/JobMonitor/ActionDto.cs
using OpenBullet2.Web.Attributes;
using RuriLib.Models.Jobs.Monitor.Actions;

namespace OpenBullet2.Web.Dtos.JobMonitor;

/// <summary>
/// Generic action DTO.
/// </summary>
public class ActionDto : PolyDto
{
}

/// <summary>
/// Waits for a certain amount of time before executing
/// the next action in the chain.
/// </summary>
[PolyType("waitAction")]
[MapsFrom(typeof(WaitAction), false)]
[MapsTo(typeof(WaitAction), false)]
public class WaitActionDto : ActionDto
{
    /// <summary>
    /// The amount of time to wait.
    /// </summary>
    public TimeSpan TimeSpan { get; set; }
}

/// <summary>
/// Sets the relative start condition of a job to the given time span.
/// </summary>
[PolyType("setRelativeStartConditionAction")]
[MapsFrom(typeof(SetRelativeStartConditionAction), false)]
[MapsTo(typeof(SetRelativeStartConditionAction), false)]
public class SetRelativeStartConditionActionDto : ActionDto
{
    /// <summary>
    /// The id of the target job.
    /// </summary>
    public int JobId { get; set; }

    /// <summary>
    /// The amount of time the target job should wait for
    /// before starting.
    /// </summary>
    public TimeSpan TimeSpan { get; set; }
}

/// <summary>
/// Stops the target job.
/// </summary>
[PolyType("stopJobAction")]
[MapsFrom(typeof(StopJobAction))]
[MapsTo(typeof(StopJobAction))]
public class StopJobActionDto : ActionDto
{
    /// <summary>
    /// The id of the target job.
    /// </summary>
    public int JobId { get; set; }
}

/// <summary>
/// Aborts the target job.
/// </summary>
[PolyType("abortJobAction")]
[MapsFrom(typeof(AbortJobAction))]
[MapsTo(typeof(AbortJobAction))]
public class AbortJobActionDto : ActionDto
{
    /// <summary>
    /// The id of the target job.
    /// </summary>
    public int JobId { get; set; }
}

/// <summary>
/// Starts the target job.
/// </summary>
[PolyType("startJobAction")]
[MapsFrom(typeof(StartJobAction))]
[MapsTo(typeof(StartJobAction))]
public class StartJobActionDto : ActionDto
{
    /// <summary>
    /// The id of the target job.
    /// </summary>
    public int JobId { get; set; }
}

/// <summary>
/// Sends a message through a discord webhook.
/// </summary>
[PolyType("discordWebhookAction")]
[MapsFrom(typeof(DiscordWebhookAction))]
[MapsTo(typeof(DiscordWebhookAction))]
public class DiscordWebhookActionDto : ActionDto
{
    /// <summary>
    /// The webhook link.
    /// </summary>
    public string Webhook { get; set; } = string.Empty;

    /// <summary>
    /// The message.
    /// </summary>
    public string Message { get; set; } = string.Empty;
}

/// <summary>
/// Sends a message through a telegram bot.
/// </summary>
[PolyType("telegramBotAction")]
[MapsFrom(typeof(TelegramBotAction))]
[MapsTo(typeof(TelegramBotAction))]
public class TelegramBotActionDto : ActionDto
{
    /// <summary>
    /// The authentication token.
    /// </summary>
    public string Token { get; set; } = string.Empty;

    /// <summary>
    /// The id of the chat.
    /// </summary>
    public long ChatId { get; set; } = 0;

    /// <summary>
    /// The message.
    /// </summary>
    public string Message { get; set; } = string.Empty;
}

File: openbullet2-web-client/src/app/main/dtos/monitor/action.dto.ts
import { TimeSpanPipe } from 'src/app/shared/pipes/timespan.pipe';
import { parseTimeSpan } from 'src/app/shared/utils/dates';

export enum ActionType {
  Wait = 'waitAction',
  SetRelativeStartCondition = 'setRelativeStartConditionAction',
  StopJob = 'stopJobAction',
  AbortJob = 'abortJobAction',
  StartJob = 'startJobAction',
  DiscordWebhook = 'discordWebhookAction',
  TelegramBot = 'telegramBotAction',
  SetBots = 'setBotsAction',
  ReloadProxies = 'reloadProxiesAction',
}

export interface WaitActionDto {
  _polyTypeName: ActionType.Wait;
  timeSpan: string;
}

export interface SetRelativeStartConditionActionDto {
  _polyTypeName: ActionType.SetRelativeStartCondition;
  jobId: number;
  timeSpan: string;
}

export interface StopJobActionDto {
  _polyTypeName: ActionType.StopJob;
  jobId: number;
}

export interface AbortJobActionDto {
  _polyTypeName: ActionType.AbortJob;
  jobId: number;
}

export interface StartJobActionDto {
  _polyTypeName: ActionType.StartJob;
  jobId: number;
}

export interface DiscordWebhookActionDto {
  _polyTypeName: ActionType.DiscordWebhook;
  webhook: string;
  message: string;
}

export interface TelegramBotActionDto {
  _polyTypeName: ActionType.TelegramBot;
  apiServer: string;
  token: string;
  chatId: number;
  message: string;
}

export interface SetBotsActionDto {
  _polyTypeName: ActionType.SetBots;
  amount: number;
}

export interface ReloadProxiesActionDto {
  _polyTypeName: ActionType.ReloadProxies;
}

export type ActionDto =
  | WaitActionDto
  | SetRelativeStartConditionActionDto
  | StopJobActionDto
  | AbortJobActionDto
  | StartJobActionDto
  | DiscordWebhookActionDto
  | TelegramBotActionDto
  | SetBotsActionDto
  | ReloadProxiesActionDto;

export function getActionText(action: ActionDto): string {
  const timeSpanPipe = new TimeSpanPipe();

  switch (action._polyTypeName) {
    case ActionType.Wait:
      return `Wait ${timeSpanPipe.transform(parseTimeSpan(action.timeSpan))}`;
    case ActionType.SetRelativeStartCondition:
      return `Set relative start condition of job ${action.jobId} to ${timeSpanPipe.transform(
        parseTimeSpan(action.timeSpan),
      )}`;
    case ActionType.StopJob:
      return `Stop job ${action.jobId}`;
    case ActionType.AbortJob:
      return `Abort job ${action.jobId}`;
    case ActionType.StartJob:
      return `Start job ${action.jobId}`;
    case ActionType.DiscordWebhook:
      return 'Send message via Discord webhook';
    case ActionType.TelegramBot:
      return 'Send message via Telegram bot';
    case ActionType.SetBots:
      return `Set bots to ${action.amount}`;
    case ActionType.ReloadProxies:
      return 'Reload proxies';
    default:
      return 'Unknown action';
  }
}
